constraint-solver/src/constraint_system.rs

```diff
@@ -0,0 +1,141 @@
+use std::hash::Hash;
+
+/// Description of a constraint system.
+pub struct ConstraintSystem<T: FieldElement, V> {
+    /// The algebraic expressions which have to evaluate to zero.
+    pub algebraic_constraints: Vec<QuadraticSymbolicExpression<T, V>>,
+    /// The bus interactions which have to be satisfied.
+    pub bus_interactions: Vec<BusInteraction<QuadraticSymbolicExpression<T, V>>>,
+}
+
+impl<T: FieldElement, V> ConstraintSystem<T, V> {
+    pub fn iter(&self) -> Box<dyn Iterator<Item = &QuadraticSymbolicExpression<T, V>> + '_> {
+        Box::new(
+            self.algebraic_constraints
+                .iter()
+                .chain(self.bus_interactions.iter().flat_map(|b| b.iter())),
+        )
+    }
+
+    pub fn iter_mut(
+        &mut self,
+    ) -> Box<dyn Iterator<Item = &mut QuadraticSymbolicExpression<T, V>> + '_> {
+        Box::new(
+            self.algebraic_constraints
+                .iter_mut()
+                .chain(self.bus_interactions.iter_mut().flat_map(|b| b.iter_mut())),
+        )
+    }
+}
+
+/// A bus interaction.
+pub struct BusInteraction<V> {
+    /// The ID of the bus.
+    pub bus_id: V,
+    /// The payload of the bus interaction.
+    pub payload: Vec<V>,
+    /// The multiplicity of the bus interaction. In most cases,
+    /// this should evaluate to 1 or -1.
+    pub multiplicity: V,
+}
+
+impl<V> BusInteraction<V> {
+    pub fn iter(&self) -> Box<dyn Iterator<Item = &V> + '_> {
+        Box::new(
+            [&self.bus_id, &self.multiplicity]
+                .into_iter()
+                .chain(self.payload.iter()),
+        )
+    }
+
+    pub fn iter_mut(&mut self) -> Box<dyn Iterator<Item = &mut V> + '_> {
+        Box::new(
+            [&mut self.bus_id, &mut self.multiplicity]
+                .into_iter()
+                .chain(self.payload.iter_mut()),
+        )
+    }
+}
+
+impl<T: FieldElement, V: Clone + Hash + Ord> BusInteraction<QuadraticSymbolicExpression<T, V>> {
+    /// Converts a bus interactions with fields represented by expressions
+    /// to a bus interaction with fields represented by range constraints.
+    pub fn to_range_constraints(
+        &self,
+        range_constraints: &impl RangeConstraintProvider<T, V>,
+    ) -> Option<BusInteraction<RangeConstraint<T>>> {
+        // TODO: Handle bus interactions with complex expressions.
+        Some(BusInteraction {
+            bus_id: expr_to_range_constraint(&self.bus_id, range_constraints)?,
+            payload: self
+                .payload
+                .iter()
+                .map(|expr| expr_to_range_constraint(expr, range_constraints))
+                .collect::<Option<Vec<_>>>()?,
+            multiplicity: expr_to_range_constraint(&self.multiplicity, range_constraints)?,
+        })
+    }
+}
+
+impl<T: FieldElement, V: Clone + Hash + Ord + Eq>
+    BusInteraction<QuadraticSymbolicExpression<T, V>>
+{
+    /// Attempts to execute the bus interaction, using the given
```

Comment:
I don't think it really executes it. It communicates with the handler to improve the constraints.